### PR TITLE
MF-513 - Add Bootstrapping service

### LINKED ISSUE #513: Add Bootstrapping service
**FEATURE REQUEST**

1. Is there an open issue addressing this request? If it does, please add a "+1" reaction to the
   existing issue, otherwise proceed to step 2.
No.

2. Describe the feature you are requesting, as well as the possible use case(s) for it.
Bootstrap service will provide initial Thing configuration. Besides this initial configuration, bootstrap service provides simple access control over Things. It’s possible to add or remove Thing to/from the whitelist in order to enable or disable message exchange for it. Initial BS endpoint will be provided in the Thing during the manufacturing process. Provisioning a new Thing is as simple as sending correspoinding data to this service. Once Thing is active, it’s aware only of the existence of that BS endpoint, so it sends a request for initial config to BS service. Once the Thing is bootstrapped, it’s possible to add it to the whitelist, so that it can exchange messages using Mainflux. Bootstrapping does not implicitly whitelist the Thing, so it has to be done manually. Thing configuration includes a list of Channels the Thing is connected to, so whitelisting is basically connecting and disconnecting the Thing to/from them. These Channels need to be preprovisioned. 
3. Indicate the importance of this feature to you (must-have, should-have, nice-to-have).
This is should-have feature.

### PR DESCRIPTION
**Pull request by dborovcanin**
### What does this do?
This pull request introduces Bootstrapping service.
### Which issue(s) does this PR fix/relate to?
This pull request resolves #513.

### Have you included tests for your changes?
Yes.

### Did you document any new/modified functionality?
Yes.

### CODE REVIEW

**Review by drasko — DISMISSED** [review]
LGTM

File: bootstrap/api/requests.go [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+
+type apiReq interface {
+	validate() error
+}
+
+type addReq struct {
+	key         string
+	ExternalID  string   `json:"external_id"`
+	ExternalKey string   `json:"external_key"`
+	Channels    []string `json:"mainflux_channels"`
```
Comment (anovakovic01) [inline bootstrap/api/requests.go]: Remove `mainflux` prefix, I think that you don't need it (correct me if I'm wrong).

File: bootstrap/configs.go [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+func (s State) String() string {
+	return strconv.Itoa(int(s))
+}
+
+const (
+	// Inactive Thing is created and configured, but not able to exchange messages using Mainflux.
+	Inactive State = iota
+	// Active Thing is created, configured, and whitelisted.
+	Active
+)
```
Comment (anovakovic01) [inline bootstrap/configs.go:15]: First define `const`, then `type` and then `func`, it's easier to follow, and also, reader of this code will be more interested in predefined values, than in implementation details.

File: bootstrap/postgres/configs.go [hunk 3]
```diff
@@ -0,0 +1,238 @@
...
+		cr.log.Error(fmt.Sprintf("Failed to retrieve configs due to %s", err))
+		return []bootstrap.Config{}
+	}
+	defer rows.Close()
+
+	items := []bootstrap.Config{}
+	var content sql.NullString
+	for rows.Next() {
+		c := bootstrap.Config{Owner: key}
+		if err = rows.Scan(&c.MFThing, &c.MFKey, pq.Array(&c.MFChannels), &c.ExternalID, &c.ExternalKey, &content, &c.State); err != nil {
```
Comment (anovakovic01) [inline bootstrap/postgres/configs.go]: Please break this in two lines, I think that this `scan` part is too long.

File: bootstrap/postgres/configs.go [hunk 4]
```diff
@@ -0,0 +1,238 @@
...
+		String: s,
+		Valid:  true,
+	}
+}
+
+func (cr configRepository) Save(thing bootstrap.Config) (string, error) {
+	q := `INSERT INTO configs (mainflux_thing, owner, mainflux_key, mainflux_channels, external_id, external_key, content, state)
+	VALUES ($1, $2, $3, $4, $5, $6, $7, $8) RETURNING mainflux_thing`
+
+	if err := cr.db.QueryRow(q, thing.MFThing, thing.Owner, thing.MFKey, pq.Array(thing.MFChannels), thing.ExternalID, thing.ExternalKey,
```
Comment (anovakovic01) [inline bootstrap/postgres/configs.go]: Please break this into two lines, this `queryrow` part is too long.

File: bootstrap/postgres/configs.go [hunk 5]
```diff
@@ -0,0 +1,238 @@
...
+
+func (cr configRepository) RetrieveByExternalID(externalKey, externalID string) (bootstrap.Config, error) {
+	q := `SELECT mainflux_thing, owner, mainflux_key, mainflux_channels, content, state FROM configs WHERE external_key = $1 AND external_id = $2`
+	var content sql.NullString
+	config := bootstrap.Config{
+		ExternalID:  externalID,
+		ExternalKey: externalKey,
+	}
+
+	if err := cr.db.QueryRow(q, externalKey, externalID).Scan(&config.MFThing, &config.Owner, &config.MFKey, pq.Array(&config.MFChannels), &content, &config.State); err != nil {
```
Comment (anovakovic01) [inline bootstrap/postgres/configs.go]: Same here, please break.

File: bootstrap/postgres/init.go [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+				Id: "configs_1",
+				Up: []string{
+					`CREATE TABLE IF NOT EXISTS configs (
+						mainflux_thing    TEXT UNIQUE NOT NULL,
+                        owner             VARCHAR(254),
+						mainflux_key      CHAR(36) UNIQUE NOT NULL,
+						mainflux_channels TEXT [],
+						external_id       TEXT UNIQUE NOT NULL,
+						external_key 	  TEXT NOT NULL,
+                        content  		  TEXT,
```
Comment (anovakovic01) [inline bootstrap/postgres/init.go]: Can you fix alignment in this file?

File: bootstrap/service.go [hunk 7]
```diff
@@ -0,0 +1,308 @@
...
+
+func (bs bootstrapService) View(key, id string) (Config, error) {
+	owner, err := bs.identify(key)
+	if err != nil {
+		return Config{}, err
+	}
+	return bs.things.RetrieveByID(owner, id)
+}
+
+func (bs bootstrapService) Update(key string, thing Config) error {
```
Comment (anovakovic01) [inline bootstrap/service.go]: This function is really long, can you break it into multiple private functions (if there are more than one logical step)?

File: cmd/bootstrap/main.go [hunk 8]
```diff
@@ -0,0 +1,215 @@
...
+	return conn
+}
+
+func startHTTPServer(svc bootstrap.Service, cfg config, logger logger.Logger, errs chan error) {
+	p := fmt.Sprintf(":%s", cfg.httpPort)
+	if cfg.serverCert != "" || cfg.serverKey != "" {
+		logger.Info(fmt.Sprintf("Bootstrap service started using https on port %s with cert %s key %s",
+			cfg.httpPort, cfg.serverCert, cfg.serverKey))
+		errs <- http.ListenAndServeTLS(p, cfg.serverCert, cfg.serverKey, api.MakeHandler(svc, bootstrap.NewConfigReader()))
+	} else {
```
Comment (anovakovic01) [inline cmd/bootstrap/main.go]: You can use `return` and remove `else`.

**Review by anovakovic01 — CHANGES_REQUESTED** [review]

File: bootstrap/postgres/init.go [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+						PRIMARY KEY (mainflux_thing, external_id)
+					)`,
+					`CREATE TABLE IF NOT EXISTS unknown (
+						external_id       TEXT UNIQUE NOT NULL,
+						external_key 	  TEXT NOT NULL,
+						PRIMARY KEY (external_id, external_key)
+					)`,
+				},
+				Down: []string{
+					"DROP TABLE things",
```
Comment (chombium) [inline bootstrap/postgres/init.go]: Drop table unknown is missing here

**Review by chombium — COMMENTED** [review]
All the endpoints which are exposed by the bootstrap service are general by their naming but are also things exclusive. Do we want to leave the `/bootstrap`, `/configs` and '/unknown' endpoints to be exclusively for things operations of we want to make the enpoints  `/things/bootstrap`, `/things/configs` and '/things/unknown' or `/bootstrap/things`, `/configs/things` and '/unknown/things', so that we leave a bit of space for eventual api upgrades and service extension to cover configuration of other Mainflux entities as well?

The other thing that bothers me is the "unknown" table name and the endpoint... /unknown... unknown what?I think that we must be more specific..

**Review by drasko — DISMISSED** [review]
LGTM

**Review by drasko — APPROVED** [review]
LGTM

**Review by anovakovic01 — APPROVED** [review]
LGTM!

**Pull request merged**